torch/_functorch/aot_autograd.py

```diff
@@ -3202,6 +3227,13 @@ def convert(idx, x):
             any(x.requires_grad for x in fake_flat_args if isinstance(x, Tensor))
             and torch.is_grad_enabled()
         )
+        if FunctionalAssertionsHelper.can_functionalize(needs_autograd=needs_autograd):
+            stack.enter_context(FunctionalAssertionsHelper.get_state_tracker())
```

Comment:
does this not work?

 ```
with FunctionalAssertionsHelper.get_state_tracker():
   ...
```

Reply:
I think the motivation here is that it only enters the scope when it can instead of always entering the scope since in current implementation: entering scope == can functionalize. How about the solution as I mentioned at https://github.com/pytorch/pytorch/pull/104203#discussion_r1244392149? Or I can have another global state to track whether it can enable asserts funcitonalization or not?